source/validate_builtins.cpp

```diff
@@ -0,0 +1,1474 @@
+{
+  // TODO(jcaraban)
+}
+
+spv_result_t CheckSecondaryPositionNV(ValidationState_t& vstate, Instruction inst,
+    uint32_t storage_class, uint32_t exec_model)
+{
+  // TODO(jcaraban)
+}
+
+spv_result_t CheckSecondaryViewportMaskNV(ValidationState_t& vstate, Instruction inst,
+    uint32_t storage_class, uint32_t exec_model)
+{
+  // TODO(jcaraban)
+}
+
+spv_result_t CheckPositionPerViewNV(ValidationState_t& vstate, Instruction inst,
+    uint32_t storage_class, uint32_t exec_model)
+{
+  // TODO(jcaraban)
+}
+
+spv_result_t CheckViewportMaskPerViewNV(ValidationState_t& vstate, Instruction inst,
+    uint32_t storage_class, uint32_t exec_model)
+{
+  // TODO(jcaraban)
+}
+*/
+
+}  // anonymous namespace
+
+namespace libspirv {
+
+#define BUILTIN_CASE(bltin) \
+  case SpvBuiltIn ## bltin: \
+    return Check ## bltin (_,inst,storage_class,execution_model);
+
+// Validates that decorations have been applied properly.
+spv_result_t ValidateBuiltIns(ValidationState_t& _) {
+  // For every instruction decorated with a built-in...
+  for (const auto& inst : _.ordered_instructions()) {
+    if (not isBuiltIn(inst,_))
+      continue;
+    auto storage_class = getStorageClass(inst);
+    auto built_in = getBuiltInEnum(inst,_);
+    // For every 'execution model' affecting the built-in, check its validty
+    for (auto execution_model : getExecutionModels(_,inst)) {
+      switch (built_in) {
+        BUILTIN_CASE( Position                    )
+        BUILTIN_CASE( PointSize                   )
+        BUILTIN_CASE( ClipDistance                )
+        BUILTIN_CASE( CullDistance                )
+        BUILTIN_CASE( VertexId                    )
+        BUILTIN_CASE( InstanceId                  )
+        BUILTIN_CASE( PrimitiveId                 )
+        BUILTIN_CASE( InvocationId                )
+        BUILTIN_CASE( Layer                       )
+        BUILTIN_CASE( ViewportIndex               )
+        BUILTIN_CASE( TessLevelOuter              )
+        BUILTIN_CASE( TessLevelInner              )
+        BUILTIN_CASE( TessCoord                   )
+        BUILTIN_CASE( PatchVertices               )
+        BUILTIN_CASE( FragCoord                   )
+        BUILTIN_CASE( PointCoord                  )
+        BUILTIN_CASE( FrontFacing                 )
+        BUILTIN_CASE( SampleId                    )
+        BUILTIN_CASE( SamplePosition              )
+        BUILTIN_CASE( SampleMask                  )
+        BUILTIN_CASE( FragDepth                   )
+        BUILTIN_CASE( HelperInvocation            )
+        BUILTIN_CASE( NumWorkgroups               )
+        BUILTIN_CASE( WorkgroupSize               )
+        BUILTIN_CASE( WorkgroupId                 )
+        BUILTIN_CASE( LocalInvocationId           )
+        BUILTIN_CASE( GlobalInvocationId          )
+        BUILTIN_CASE( LocalInvocationIndex        )
+        // TODO(jcaraban): the built-ins below are not clearly documented
```

Comment:
Not all builtins are used by Vulkan.    If the Vulkan environment spec doesn't describe it then it's not allowed in Vulkan.